Build viable Docker images for Polaris using Quarkus

adutra opened the pull request:
This PR redesigns how Docker images are built, to make them fully compatible with Quarkus. With Quarkus, docker images can be built as part of the normal build. There is no need to do it manually.

This PR also fixes the issues described in #537.

## Summary of changes

### Docker images

* The Dockerfile at the root of the repository has been removed (not needed anymore, and has security flaws – see #537 for details);
* The main Dockerfile for Polaris server is now located at `quarkus/server/src/main/docker/Dockerfile.jvm` (default location for dockerfiles for Quarkus)
* The main Dockerfile for Polaris admin tool is now located at `quarkus/admin/src/main/docker/Dockerfile.jvm`
* Both docker images can be built during the regular build: `./gradlew assemble -Dquarkus.container-image.build=true`
* The images are built with the Quarkus `docker` build provider. You must have a valid docker environment for the image build to succeed.

<strike>
### JDBC drivers

Furthermore, this PR modifies the eclipse-link module and includes 2 JDBC drivers in the resulting artifacts: H2 and Postgres. 

**This is absolutely required to be able to run Polaris with Postgres as the JDBC drivers _must_ be present at build time.**

The licenses of the 2 drivers are compatible with ASF and were included in the relevant files.
</strike>

### Helm tests

This PR changes the Helm chart _slightly_, only to make the Helm tests pass in CI. But another PR will introduce a bigger revamp of the Helm chart in order to fully to adapt it to Quarkus: #626.

In CI, a docker image is first built with Gradle and Quarkus, then loaded into minikube. Then the tests are executed normally.

### Regression tests

* The docker-compose.yaml at the root of the repository is moved to the `regtests` folder where it naturally belongs.
* The default image in the `docker-compose.yaml` is `apache/polaris:latest`. This is the image name that is created when running `./gradlew assemble -Dquarkus.container-image.build=true`.
* In CI, a docker image is first built with Gradle and Quarkus. Then the tests are executed.

Note: there is no released version of Polaris yet, so running the docker compose file today will result in an error unless the user builds the image manually first with  `./gradlew assemble -Dquarkus.container-image.build=true`; but that's a temporary issue.

## Remaining work

Again, this PR also defers all documentation changes to a follow-up task: #700

Also, this PR does not address the specific needs of release builds. For such builds, we need:

* Multi-platform manifests
* Multi-registry pushes

adutra commented:
Ready for review.

dimas-b commented on quarkus/service/src/main/resources/application.properties:
```diff
@@ -82,7 +82,9 @@ polaris.features.defaults."SUPPORTED_CATALOG_STORAGE_TYPES"=["S3","GCS","AZURE",
 # polaris.features.realm-overrides."my-realm"."SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION"=true
 
-polaris.persistence.type=in-memory
+polaris.persistence.type=eclipse-link
```
Not sure about using `eclipse-link` by default. We can always set it in Helm values, but running from the jar locally and having to override it back to in-memory (most of the time) seems inconvenient.

adutra replied:
See my previous comment. 

You shouldn't need to switch to `in-memory` when running locally. H2 is configured with an in-memory database, so in the end, it's the same from a UX perspective.

dimas-b replied:
Ah, I missed H2 in-memory... I assumed it was file-based :facepalm:

eric-maynard replied:
H2 in-memory and the TreeMap metastore do not have the same semantics. For example, the in-memory metastore currently prints credentials when it bootstraps.

adutra replied:
Exactly, and you don't want your credentials printed when running in production right? These settings are the production ones.

For running locally, we can instruct users to switch to in-memory:

```shell
 docker run -it -p 8181:8181 -e polaris.persistence.type=in-memory apache/polaris
```

Then the credentials get printed to the console.

dimas-b reviewed (commented):
LGTM overall. Thanks @adutra ! Just a few minor comments.

adutra commented on extension/persistence/eclipselink/build.gradle.kts:
```diff
@@ -65,6 +65,10 @@ dependencies {
   compileOnly("com.fasterxml.jackson.core:jackson-core")
 
+  // JDBC Drivers
+  runtimeOnly(libs.h2)
```
So, I kept H2 for now because I also changed the default persistence type to `eclipse-link`. I think we need to achieve 2 goals here:

1. The default configuration should be as "production-ready" as possible
2. The default configuration should make Polaris start successfully.

The 2 goals are sometimes in conflict. I think that by switching to `eclipse-link` by default we achieve goal 1), and by using H2 by default, we achieve 2) – even if H2 is not really "production ready", that's the best we can do.

In any case, the `in-memory` persistence type imho should now be reserved only for tests. We could (should?) even rename it to `test`.

dimas-b replied:
Fair point :+1:

eric-maynard commented:
@MonkeyCanCode -- might be worth taking a look at this if you have time

eric-maynard commented on helm/polaris/values.yaml:
```diff
@@ -23,7 +23,7 @@ replicaCount: 1
 image:
   # -- The image repository to pull from.
-  repository: localhost:5001/polaris
+  repository: apache/polaris
```
This is different, right?

dimas-b replied:
Different, but I think `apache/polaris` is more correct :)

adutra replied:
Exactly, `localhost:5001` is definitely not a good choice. This is only useful if you have a local registry running, but images with that kind of registry are otherwise unusable.

Since @jbonofre created our DockerHub space: https://hub.docker.com/r/apache/polaris, the official polaris image name will be `apache/polaris` - which expands to `docker.io/apache/polaris`.

This is the name I'm using consistently in this PR. E.g. if you build the images as of today with `./gradlew assemble -Dquarkus.container-image.build=true` you would get the following images loaded in your Docker environment:

* `apache/polaris:latest`
* `apache/polaris:1.0.0-incubating-SNAPSHOT`
* `apache/polaris-admin-tool:latest`
* `apache/polaris-admin-tool:1.0.0-incubating-SNAPSHOT`

eric-maynard commented on quarkus/admin/build.gradle.kts:
```diff
@@ -35,6 +35,7 @@ dependencies {
   implementation(project(":polaris-api-iceberg-service"))
   implementation(project(":polaris-service-common"))
   implementation(project(":polaris-quarkus-service"))
+  runtimeOnly(project(":polaris-eclipselink"))
```
Is this needed now because you're switching the default metastore?

adutra replied:
No, it's needed because all the dependencies we need must be included _in the build_: you cannot add this dependency when deploying Polaris. Even if the default metastore was `foo`, if we want to support the `eclipse-link` metastore, all the eclipselink dependencies must be included in the final artifact (and those of `foo`).

Maybe the comparison with Nessie could help here: Nessie supports H2, Postgres, MySQL, MariaDB, Cassandra, MongoDB, DynamoDB, and BigTable. Therefore, all the drivers for these backends are included in Nessie builds. Granted, this increases the artifact size, but on the other hand, the artifact "just works" for everybody. And to be honest, Java applications have never been famous for their small sizes :-)

eric-maynard replied:
> if we want to support the eclipse-link metastore, all the eclipselink dependencies must be included in the final artifact

Previously, it was my understanding that we asked users to supply additional dependencies as command-line arguments.

> Nessie supports H2, Postgres, MySQL, ...

By supporting EclipseLink, Polaris supports innumerable databases. Surely we won't include them all here?

adutra replied:
I know it's a bit of a paradigm shift, but we get used to it quickly: adding dependencies at run time is not possible with frameworks like Quarkus. They must be added at build time instead.

We won't include all possible JDBC drivers, but we can (and should) include the most popular ones (if their licenses are compatible). I think Postgres is by far the one that should definitely be included. But we can include many others if there is reasonable ask for that from the community. Nessie ships with something like 8 drivers, and that's fine. But if someone wants to use Polaris (or Nessie) against a driver that is not present by default – say, Oracle – then they need to rebuild the application themselves and include the extra dependencies in their own build.

Hope that's a bit more clear now :-)

eric-maynard replied:
Got it, I didn't realize that adding dependencies at runtime wasn't possible with Quarkus. That seems like something we probably should have discussed before switching over to Quarkus.

jbonofre replied:
@eric-maynard yes and no 😄 Quarkus does the scanning at build time, that's true. It's the fastest and easiest way. It's possible to load a dependencies wrap to be loaded at runtime (with Extension/Qualifier and creating the jandex index, etc) but it requires additional hook (like custom URLClassLoader, e.g. Quarkus community is discussing about "extended" classloader manager https://github.com/quarkusio/quarkus/discussions/43749).

For now, I suggest to keep the "regular" and simple approach as @adutra proposed. We can improve the loading later (with meta extension/classloader manager).

eric-maynard replied:
In the spirit of keeping main in a working state, perhaps we should first try to restore the behavior pre-Quarkus (or something close to it) and then we can later follow up with improvement?

If that would take significantly more time, then I would be OK proceeding with the minimal set of changes here needed to restore working order.

dimas-b replied:
I guess this change can be in a separate PR now that the docker image does not use EclipseLink.

eric-maynard reviewed (commented):
In general the changes look good to me, but it seems that we are sneaking in a few changes unrelated to just building a viable Docker image here. Namely, we're changing the default metastore and the JRE.

MonkeyCanCode commented:
> @MonkeyCanCode -- might be worth taking a look at this if you have time

Sure thing. Will check later tonight.

adutra commented:
> In general the changes look good to me, but it seems that we are sneaking in a few changes unrelated to just building a viable Docker image here. Namely, we're changing the default metastore and the JRE.

Thanks for the review @eric-maynard  ! If you prefer I can revert the default metastore change.

However, what do you mean by changing the JRE? It has always been `openjdk-21`.

adutra commented:
OK, I reverted the default persistence type back to `in-memory` as I don't want to hold this PR for longer than strictly required.

cc @dimas-b @eric-maynard

eric-maynard commented on .github/workflows/helm.yml:
```diff
@@ -43,6 +43,12 @@ jobs:
+        uses: actions/setup-java@v4
+        with:
+          java-version: '21'
+          distribution: 'temurin'
```
If I'm not mistaken this is a different JDK than the one we used previously

eric-maynard commented on extension/persistence/eclipselink/build.gradle.kts:
```diff
@@ -65,6 +65,10 @@ dependencies {
 
+  // JDBC Drivers
+  runtimeOnly(libs.h2)
+  runtimeOnly(libs.postgresql)
```
If you're going back to in-memory by default, these changes aren't needed either

dimas-b replied:
I think it is valuable to end users that Polaris's normal build includes usable JDBC drivers. I do not think it is viable to force users to recompile to include PostgreSQL.

eric-maynard commented on extension/persistence/eclipselink/src/main/resources/META-INF/persistence.xml:
```diff
@@ -34,8 +34,7 @@
     <shared-cache-mode>NONE</shared-cache-mode>
     <properties>
-      <property name="jakarta.persistence.jdbc.url"
-        value="jdbc:h2:file:./build/test_data/polaris/{realm}/db"/>
```
Why is this changing?

adutra replied:
Because I don't see the value of creating a file on disk? 
That said, if you really insist on having that file created, I can revert this change.

snazy replied:
I agree - using in-memory here is better.

eric-maynard commented on gradle/libs.versions.toml:
```diff
@@ -81,6 +81,7 @@ opentelemetry-bom = { module = "io.opentelemetry:opentelemetry-bom", version = "
 opentelemetry-semconv = { module = "io.opentelemetry.semconv:opentelemetry-semconv", version = "1.25.0-alpha" }
 picocli = { module = "info.picocli:picocli-codegen", version.ref = "picocli" }
 picocli-codegen = { module = "info.picocli:picocli-codegen", version.ref = "picocli" }
+postgresql = { module = "org.postgresql:postgresql", version = "42.7.4" }
```
Not needed

adutra replied:
Again, this is needed. Otherwise the resulting Docker images will be unusable.

adutra replied:
I mean, we can remove this and ship just with H2 for now. But I don't see the point of a Polaris Docker image that is only usable with H2.

dimas-b replied:
This declaration does look now (after recent review-driven changes)

jbonofre commented:
@eric-maynard just for my understanding, you are asking to include changes only related to Docker here (and not include changes on `persistence.xml`) right ? This is the change you are requesting: revert the change on `persistence.xml`, correct ?

snazy reviewed (approved):
The change looks good to me and comply with what's needed for Quarkus.
I don't see a technical reason to block this change.

@eric-maynard please retract your review or provide the obligatory justified technical reason.

eric-maynard commented:
@jbonofre 
> @eric-maynard just for my understanding, you are asking to include changes only related to Docker here (and not include changes on persistence.xml) right ? This is the change you are requesting: revert the change on persistence.xml, correct ?

For now, yes. I think we should try to first restore the docker workflow to its previous state or something close to it before making improvements to the metastore that are not strictly related to building a Docker image.

dimas-b commented on quarkus/service/src/main/java/org/apache/polaris/service/quarkus/config/QuarkusProducers.java:
```diff
@@ -181,6 +182,10 @@ public void maybeInitializeInMemoryRealm(
     if (factory instanceof InMemoryPolarisMetaStoreManagerFactory) {
+      LoggerFactory.getLogger(QuarkusProducers.class)
+          .warn(
+              "Polaris is configured with an in-memory metastore; "
```
nit: maybe move to a separate PR for clarity?

adutra commented:
> This, however, makes are few other changes in this PR unnecessary, so it might be best to move them to a separate PR for the sake of clarity. 

Indeed this PR needs some more cleanup. Working on it.

adutra commented on quarkus/admin/build.gradle.kts:
```diff
@@ -37,6 +37,10 @@ dependencies {
   implementation(project(":polaris-quarkus-service"))
   implementation(project(":polaris-quarkus-defaults"))
 
+  if (project.hasProperty("eclipseLinkDeps")) {
```
Since we are not including any JDBC drivers anymore; we now need this hook.

dimas-b replied:
Looks like a fair trade-off to me.

adutra commented:
OK I think I've cleaned up everything, removed JDBC drivers, removed unrelated changes. The scope is now considerably smaller.

@dimas-b @eric-maynard please have a second look.

dimas-b reviewed (approved):
LGTM :+1:

adutra commented:
I would like to merge this PR today, if that's OK for everybody. @eric-maynard please review and if fine with you, remove your -1.

eric-maynard reviewed (approved):
LGTM; the `eclipseLinkDeps` change makes sense to me
